File: app/src/main/java/com/xnote/app/domain/document/InlineEditing.kt
package com.xnote.app.domain.document

// -- Type Definitions

data class InlineMarks(
    val bold: Boolean = false,
    val italic: Boolean = false,
    val underline: Boolean = false,
    val strikethrough: Boolean = false,
    val highlight: Boolean = false,
    val linkUrl: String? = null,
) {
    fun toRun(text: String): InlineRun = InlineRun(
        text = text,
        bold = bold,
        italic = italic,
        underline = underline,
        strikethrough = strikethrough,
        highlight = highlight,
        linkUrl = linkUrl,
    )
}

enum class InlineMark {
    Bold,
    Italic,
    Underline,
    Strikethrough,
    Highlight,
}

// -- Functions

fun InlineRun.marks(): InlineMarks = InlineMarks(
    bold = bold,
    italic = italic,
    underline = underline,
    strikethrough = strikethrough,
    highlight = highlight,
    linkUrl = linkUrl,
)

fun InlineRun.withMarks(marks: InlineMarks): InlineRun = marks.toRun(text)

fun InlineRun.sameStyle(other: InlineRun): Boolean = marks() == other.marks()

fun InlineMarks.has(mark: InlineMark): Boolean = when (mark) {
    InlineMark.Bold -> bold
    InlineMark.Italic -> italic
    InlineMark.Underline -> underline
    InlineMark.Strikethrough -> strikethrough
    InlineMark.Highlight -> highlight
}

fun InlineMarks.toggle(mark: InlineMark, enable: Boolean): InlineMarks = when (mark) {
    InlineMark.Bold -> copy(bold = enable)
    InlineMark.Italic -> copy(italic = enable)
    InlineMark.Underline -> copy(underline = enable)
    InlineMark.Strikethrough -> copy(strikethrough = enable)
    InlineMark.Highlight -> copy(highlight = enable)
}

fun List<InlineRun>.coalesce(): List<InlineRun> {
    if (isEmpty()) return emptyList()
    val merged = ArrayList<InlineRun>(size)
    for (run in this) {
        if (run.text.isEmpty()) continue
        val last = merged.lastOrNull()
        if (last != null && last.sameStyle(run)) {
            merged[merged.lastIndex] = last.copy(text = last.text + run.text)
        } else {
            merged += run
        }
    }
    return merged
}

fun List<InlineRun>.splitAt(offset: Int): Pair<List<InlineRun>, List<InlineRun>> {
    val length = plainText().length
    val safe = offset.coerceIn(0, length)
    if (safe <= 0) return emptyList<InlineRun>() to coalesce()
    if (safe >= length) return coalesce() to emptyList()

    val before = ArrayList<InlineRun>()
    val after = ArrayList<InlineRun>()
    var remaining = safe
    var split = false
    for (run in this) {
        if (split) {
            after += run
            continue
        }
        val runLength = run.text.length
        when {
            remaining >= runLength -> {
                before += run
                remaining -= runLength
                if (remaining == 0) split = true
            }
            remaining == 0 -> {
                after += run
                split = true
            }
            else -> {
                before += run.copy(text = run.text.substring(0, remaining))
                after += run.copy(text = run.text.substring(remaining))
                split = true
            }
        }
    }
    return before.coalesce() to after.coalesce()
}

fun List<InlineRun>.replaceRange(start: Int, end: Int, insertion: InlineRun): List<InlineRun> {
    val length = plainText().length
    val lo = minOf(start, end).coerceIn(0, length)
    val hi = maxOf(start, end).coerceIn(0, length)
    val (left, rest) = splitAt(lo)
    val (_, right) = rest.splitAt(hi - lo)
    val middle = if (insertion.text.isEmpty()) emptyList() else listOf(insertion)
    return (left + middle + right).coalesce()
}

fun List<InlineRun>.mapRange(start: Int, end: Int, transform: (InlineRun) -> InlineRun): List<InlineRun> {
    val length = plainText().length
    val lo = minOf(start, end).coerceIn(0, length)
    val hi = maxOf(start, end).coerceIn(0, length)
    if (lo == hi) return coalesce()
    val (left, rest) = splitAt(lo)
    val (middle, right) = rest.splitAt(hi - lo)
    return (left + middle.map(transform) + right).coalesce()
}

fun List<InlineRun>.runsInRange(start: Int, end: Int): List<InlineRun> {
    val length = plainText().length
    val lo = minOf(start, end).coerceIn(0, length)
    val hi = maxOf(start, end).coerceIn(0, length)
    if (lo == hi) return emptyList()
    val (_, rest) = splitAt(lo)
    val (middle, _) = rest.splitAt(hi - lo)
    return middle
}

fun List<InlineRun>.marksAt(offset: Int): InlineMarks {
    if (isEmpty()) return InlineMarks()
    val length = plainText().length
    if (length == 0) return first().marks()
    val index = if (offset <= 0) 0 else (offset - 1).coerceIn(0, length - 1)
    var cursor = 0
    for (run in this) {
        val next = cursor + run.text.length
        if (index < next) return run.marks()
        cursor = next
    }
    return last().marks()
}

fun List<InlineRun>.rangeHasMark(start: Int, end: Int, mark: InlineMark): Boolean {
    val runs = runsInRange(start, end)
    return runs.isNotEmpty() && runs.all { it.marks().has(mark) }
}

fun List<InlineRun>.rangeHasLink(start: Int, end: Int): Boolean {
    val runs = runsInRange(start, end)
    return runs.isNotEmpty() && runs.all { !it.linkUrl.isNullOrBlank() }
}

fun findTextReplacement(oldText: String, newText: String): Triple<Int, Int, String> {
    if (oldText == newText) {
        return Triple(newText.length, newText.length, "")
    }
    var start = 0
    val shared = minOf(oldText.length, newText.length)
    while (start < shared && oldText[start] == newText[start]) {
        start += 1
    }
    var oldEnd = oldText.length
    var newEnd = newText.length
    while (oldEnd > start && newEnd > start && oldText[oldEnd - 1] == newText[newEnd - 1]) {
        oldEnd -= 1
        newEnd -= 1
    }
    return Triple(start, oldEnd, newText.substring(start, newEnd))
}

File: flutter/lib/domain/document/inline_editing.dart
import 'note_block.dart';

// -- Type Definitions

enum InlineMark { bold, italic, underline, strikethrough, highlight }

final class InlineMarks {
  const InlineMarks({
    this.bold = false,
    this.italic = false,
    this.underline = false,
    this.strikethrough = false,
    this.highlight = false,
    this.linkUrl,
  });

  final bool bold;
  final bool italic;
  final bool underline;
  final bool strikethrough;
  final bool highlight;
  final String? linkUrl;

  // -- Functions

  InlineRun toRun(String text) => InlineRun(
        text: text,
        bold: bold,
        italic: italic,
        underline: underline,
        strikethrough: strikethrough,
        highlight: highlight,
        linkUrl: linkUrl,
      );

  bool has(InlineMark mark) => switch (mark) {
        InlineMark.bold => bold,
        InlineMark.italic => italic,
        InlineMark.underline => underline,
        InlineMark.strikethrough => strikethrough,
        InlineMark.highlight => highlight,
      };

  InlineMarks toggle(InlineMark mark, {required bool enabled}) {
    return InlineMarks(
      bold: mark == InlineMark.bold ? enabled : bold,
      italic: mark == InlineMark.italic ? enabled : italic,
      underline: mark == InlineMark.underline ? enabled : underline,
      strikethrough: mark == InlineMark.strikethrough ? enabled : strikethrough,
      highlight: mark == InlineMark.highlight ? enabled : highlight,
      linkUrl: linkUrl,
    );
  }

  InlineMarks copyWith({Object? linkUrl = _notProvided}) {
    return InlineMarks(
      bold: bold,
      italic: italic,
      underline: underline,
      strikethrough: strikethrough,
      highlight: highlight,
      linkUrl:
          identical(linkUrl, _notProvided) ? this.linkUrl : linkUrl as String?,
    );
  }

  @override
  bool operator ==(Object other) {
    return other is InlineMarks &&
        other.bold == bold &&
        other.italic == italic &&
        other.underline == underline &&
        other.strikethrough == strikethrough &&
        other.highlight == highlight &&
        other.linkUrl == linkUrl;
  }

  @override
  int get hashCode => Object.hash(
        bold,
        italic,
        underline,
        strikethrough,
        highlight,
        linkUrl,
      );
}

final class TextReplacement {
  const TextReplacement({
    required this.start,
    required this.end,
    required this.insertedText,
  });

  final int start;
  final int end;
  final String insertedText;
}

// -- Constants

const Object _notProvided = Object();

// -- Functions

InlineMarks marksOf(InlineRun run) => InlineMarks(
      bold: run.bold,
      italic: run.italic,
      underline: run.underline,
      strikethrough: run.strikethrough,
      highlight: run.highlight,
      linkUrl: run.linkUrl,
    );

InlineRun withMarks(InlineRun run, InlineMarks marks) => marks.toRun(run.text);

extension InlineRunListEditing on List<InlineRun> {
  List<InlineRun> coalesced() {
    final merged = <InlineRun>[];
    for (final run in this) {
      if (run.text.isEmpty) {
        continue;
      }
      final last = merged.lastOrNull;
      if (last != null && marksOf(last) == marksOf(run)) {
        merged[merged.length - 1] =
            last.copyWith(text: '${last.text}${run.text}');
      } else {
        merged.add(run);
      }
    }
    return List<InlineRun>.unmodifiable(merged);
  }

  (List<InlineRun>, List<InlineRun>) splitAt(int offset) {
    final safeOffset = offset.clamp(0, plainText(this).length);
    if (safeOffset == 0) {
      return (<InlineRun>[], coalesced());
    }
    if (safeOffset == plainText(this).length) {
      return (coalesced(), <InlineRun>[]);
    }

    final before = <InlineRun>[];
    final after = <InlineRun>[];
    var remaining = safeOffset;
    var didSplit = false;
    for (final run in this) {
      if (didSplit) {
        after.add(run);
        continue;
      }
      if (remaining >= run.text.length) {
        before.add(run);
        remaining -= run.text.length;
        didSplit = remaining == 0;
      } else if (remaining == 0) {
        after.add(run);
        didSplit = true;
      } else {
        before.add(run.copyWith(text: run.text.substring(0, remaining)));
        after.add(run.copyWith(text: run.text.substring(remaining)));
        didSplit = true;
      }
    }
    return (before.coalesced(), after.coalesced());
  }

  List<InlineRun> replacedRange(int start, int end, InlineRun insertion) {
    final length = plainText(this).length;
    final lower = (start < end ? start : end).clamp(0, length);
    final upper = (start > end ? start : end).clamp(0, length);
    final (left, rest) = splitAt(lower);
    final (_, right) = rest.splitAt(upper - lower);
    final middle =
        insertion.text.isEmpty ? <InlineRun>[] : <InlineRun>[insertion];
    return <InlineRun>[...left, ...middle, ...right].coalesced();
  }

  List<InlineRun> mapRange(
    int start,
    int end,
    InlineRun Function(InlineRun run) transform,
  ) {
    final length = plainText(this).length;
    final lower = (start < end ? start : end).clamp(0, length);
    final upper = (start > end ? start : end).clamp(0, length);
    if (lower == upper) {
      return coalesced();
    }
    final (left, rest) = splitAt(lower);
    final (middle, right) = rest.splitAt(upper - lower);
    return <InlineRun>[
      ...left,
      ...middle.map(transform),
      ...right,
    ].coalesced();
  }

  List<InlineRun> runsInRange(int start, int end) {
    final length = plainText(this).length;
    final lower = (start < end ? start : end).clamp(0, length);
    final upper = (start > end ? start : end).clamp(0, length);
    if (lower == upper) {
      return <InlineRun>[];
    }
    final (_, rest) = splitAt(lower);
    final (middle, _) = rest.splitAt(upper - lower);
    return middle;
  }

  InlineMarks marksAt(int offset) {
    if (isEmpty) {
      return const InlineMarks();
    }
    final length = plainText(this).length;
    if (length == 0) {
      return marksOf(first);
    }
    final index = offset <= 0 ? 0 : (offset - 1).clamp(0, length - 1);
    var cursor = 0;
    for (final run in this) {
      final next = cursor + run.text.length;
      if (index < next) {
        return marksOf(run);
      }
      cursor = next;
    }
    return marksOf(last);
  }

  bool rangeHasMark(int start, int end, InlineMark mark) {
    final runs = runsInRange(start, end);
    return runs.isNotEmpty && runs.every((run) => marksOf(run).has(mark));
  }

  bool rangeHasLink(int start, int end) {
    final runs = runsInRange(start, end);
    return runs.isNotEmpty &&
        runs.every((run) => run.linkUrl?.trim().isNotEmpty ?? false);
  }
}

TextReplacement findTextReplacement(String oldText, String newText) {
  if (oldText == newText) {
    return TextReplacement(
      start: newText.length,
      end: newText.length,
      insertedText: '',
    );
  }
  var start = 0;
  final sharedLength =
      oldText.length < newText.length ? oldText.length : newText.length;
  while (start < sharedLength && oldText[start] == newText[start]) {
    start += 1;
  }
  var oldEnd = oldText.length;
  var newEnd = newText.length;
  while (oldEnd > start &&
      newEnd > start &&
      oldText[oldEnd - 1] == newText[newEnd - 1]) {
    oldEnd -= 1;
    newEnd -= 1;
  }
  return TextReplacement(
    start: start,
    end: oldEnd,
    insertedText: newText.substring(start, newEnd),
  );
}
